### ISSUE TITLE
GetExternalIPAddress returns NAT-ed address from private IPv4 block

### ISSUE THREAD

**Issue by pali** [miniupnp/miniupnp#298]
It is very often that user's router with miniupnpd is already behind NAT.

Lot of ISPs provides public IPv4 address to end user just via NAT 1:1 mechanism. Which means that DHCPv4 server assigns for user router one private IPv4 address and ISP configure his infrastructure to translate user's public IPv4 address to that private IPv4 address assigned by DHCPv4 server (it is like forwarding all TCP and UDP ports).

So in reality user has his own public IPv4 address (in some cases it is just dynamic and may change in next day or weeks), but with NAT 1:1.

And in this case miniupnpd function GetExternalIPAddress() returns that private IPv4 address instead of correct public. Software which uses UPnP protocol for getting public IPv4 address does not work.

I would suggest to extend miniupnpd functionality to check if ip address on external interface is really public IPv4 (not in 10.0.0.0/8, 192.168.0.0/24, ... blocks) and configuration option to learn correct public IPv4 address via e.g. STUN protocol from STUN server. Via STUN protocol it is also possible to detect if user has above NAT 1:1 or symmetric NAT.

Maybe in case when symmetric NAT is detected miniupnpd should reject any request for ip address forwarding and return error (as it would not work).

**Comment by pali** [comment]
@miniupnp ISP forwards all tcp and udp ports from public ip address to private address assigned by dhcp. So there is no NAT traversal, it is NAT 1:1.

This setup is very common and the only thing which is needed is to learn correct public ip address.

**Comment by miniupnp** [comment]
what's the point in using NAT 1:1 for an ISP ?
usually ISP which have not enough public IP address available use RFC1918 address for their customers and use NAT to map (for example) 256 public IP addresses on 256*16=4096 customers or whatever.

In the case of NAT 1:1, why don't you use the `ext_ip=` option in miniupnpd.conf?

**Comment by pali** [comment]
> what's the point in using NAT 1:1 for an ISP ?

I'm not ISP, so I do not know exactly. I think that it simplify configuration for ISP's DHCP server which assigns IP addresses when ISP has customers with and also without public IP address. For customer which ask for public IP address, ISP just add a new record to their database and do not need to change routing nor how DHCP server assign IP addresses.

E.g. in Slovakia this is the most common behavior when end user ask for public IP address. I have never seen in Slovakia that some ISP's DHCP server assigned directly public IP address. Always just some 10.x.x.x address and tell you that you have your own public IP address with NAT 1:1 and you can run normally server applications as you want.

And another common thing: you get public IP address, but not static. Once you release your DHCP lease and ask for a new address then ISP assign for you a new public IP address (NATed 1:1; yes free IP address just for you until your DHCP lease is active). And if you want static public IP address you need to pay additional cost.

> usually ISP which have not enough public IP address available use RFC1918 address for their customers and use NAT to map (for example) 256 public IP addresses on 256*16=4096 customers or whatever.

Yes, ISPs which do not have lot of public IP addresses, just put more users behind one public IP address. But there are still ISPs which have enough IPv4 addresses and due to this fact they do not offer IPv6 at all.

> In the case of NAT 1:1, why don't you use the ext_ip= option in miniupnpd.conf?

This is working fine if you know that IP address. And also you need to hardcode into config file. Personally I do not like hardcoding any address or other configuration option when there is a way how to obtain this option.

In case that public IP address changes at time of obtaining DHCP lease, you need to get somehow current IP address, so cannot be hardcoded.

Therefore to simplify setup, I would suggest extension to miniupnd to obtain public IP address via STUN.

**Comment by miniupnp** [comment]
So in order to avoid using the ext_ip= (or -o on command line) option, you suggest that miniupnpd uses STUN to obtain public IP address. But to use STUN we need to hardcode a STUN server address !

**Comment by pali** [comment]
In case public ip address is possible to obtain only after you have dhcp lease, then you cannot create config file before dhcp client finish handshake. Which is really not usable. And you would need to use some shell magic (e.g. sed or whatever) to put obtained address into that config file.

I suggested to use STUN server. You can put some default into config file or let user to configure STUN server (and have no default there). Difference is that address of STUN server is not chaning and lot of users can use it. But public IP address is something which every user have different.

Reason why I suggested STUN server is because STUN protocol tells you if you are behind NAT 1:1 and miniupnpd would work fine. OR you are behind symetric NAT and miniupnpd would not work at all (unless upstream link supports UPnP or PCP too).

Basically it would allow you to create one configuration (with one working STUN server) which would work in many different networks.

Having ext_ip= in configuration would mean that configuration file is specific to just one network and so configured router would not work in different network.

I hope it is more clean now what I mean.

**Comment by gizahNL** [comment]
Would it not make more sense to write your own cron script that does some stun foo and updates the config file and sends miniupnpd a HUP as needed? 

Feels a lot more KISS imho

**Comment by pali** [comment]
As such networks are common, I thought that it would make sense to have such option directly in miniupnpd instead of creating a new software which would be integrated into system... Therefore I wrote request on github miniupnpd project page.

Yes, custom scripts, started from cron could work. But this is not something for end users.

As @miniupnp wrote, another common behavior is when ISPs have not enough public IP address and put users behind symmetric (or other) NAT.

Currently miniupnpd is used in lot of (chineese) devices, including famous OpenWRT system. Therefore there are lot of users of it and more times it is running on router which is behand NAT.
Proposed STUN protocol is able to detect such networks and would allows miniupnpd to reject port forwarding requests in such networks.

**Comment by miniupnp** [comment]
@pali you may not have noticed that miniupnpd is free software and that you can implement whatever feature you want.
many home routers device include a modified version of miniupnpd.

**Comment by pali** [comment]
I have noticed that it is free software... But before I decide if I would start implementing such thing... would you accept such feature? Or closed status in this ticket means "such funcionality is rejected". Based on this I would decide if I create my own set of scripts and external applications around miniupnd, or I could start implemening it as I descibed... Please do not take me wrong, I'm just trying to discuss about this funcionality and trying to describe why it can be useful.

**Comment by miniupnp** [comment]
It works the other way around :
Somebody implements a feature for a specific need. Then they send the patch upstream for merging.

**Comment by pali** [comment]
So would you accept a patch which implements this funcionality? Or it is something which you do not want in upstream miniupnpd?

**Comment by miniupnp** [comment]
It depends on several parameters.
Sometimes, the merge is not possible. For example the "ix2015" is a branch which was never merged with the master branch : https://github.com/miniupnp/miniupnp/blob/ix2015/miniupnpd/ix2015/README.txt

**Comment by miniupnp** [comment]
anyway you're asking the wrong questions at the wrong time. If you really need a feature, just implement it. It would then be the time to discuss about merging upstream.
Note : miniupnpd is a project with more than 10 years of life, and you are the only one that ever asked for this feature, so let me doubt there is much people interested.

**Comment by pali** [comment]
I published my implementation in this pull request: https://github.com/miniupnp/miniupnp/pull/307

**Comment by miniupnp** [comment]
thanks

**Comment by pali** [comment]
if you are going to review/merge it and there is something needed to rework, let me know. I will try to fix/change it.

**Comment by miniupnp** [comment]
I'm a bit busy theses days and your pull request need to be reviewed seriously

**Comment by pali** [comment]
Ok, so when you have a time, please look at it.
